src/object.hpp

```diff
@@ -1172,31 +1231,102 @@ borrowed_object writable_object<Derived>::emplace(std::string_view key, T &&valu
     return emplace(key, owned_object{std::forward<T>(value)});
 }
 
-inline borrowed_object::borrowed_object(owned_object &obj) : obj_(obj.ptr()) {}
+inline borrowed_object::borrowed_object(owned_object &obj) : obj_(obj.ptr()), alloc_(obj.alloc()) {}
 
 // NOLINTNEXTLINE(cppcoreguidelines-rvalue-reference-param-not-moved)
 inline borrowed_object &borrowed_object::operator=(owned_object &&obj)
 {
-    ref() = obj.move();
+    alloc_ = obj.alloc();
+    *obj_ = obj.move();
     return *this;
 }
+namespace object_builder {
 
-inline owned_object owned_object::make_array(
-    std::initializer_list<detail::initializer::movable_object> list)
+struct movable_object {
+    // NOLINTNEXTLINE(google-explicit-constructor,hicpp-explicit-conversions)
+    template <typename T> movable_object(T &&value) : object{std::forward<T>(value)} {}
+    movable_object(const movable_object &) = delete;
+    movable_object(movable_object &&) = delete;
+    movable_object operator=(const movable_object &) = delete;
+    movable_object operator=(movable_object &&) = delete;
+    ~movable_object() = default;
+    mutable owned_object object;
+};
+
+using all_types = std::variant<bool, int16_t, int32_t, int64_t, uint16_t, uint32_t, uint64_t,
+    double, const char *, std::string_view, std::string, movable_object>;
+using key_value = std::pair<std::string_view, all_types>;
+
+inline owned_object array(std::initializer_list<all_types> list = {},
+    memory::memory_resource *alloc = memory::get_default_resource())
 {
-    auto container = make_array();
-    for (const auto &value : list) { container.emplace_back(std::move(value.object)); }
+    auto container = owned_object::make_array(static_cast<uint16_t>(list.size()), alloc);
+    for (const auto &value : list) {
+        if (std::holds_alternative<bool>(value)) {
+            container.emplace_back(owned_object{std::get<bool>(value)});
```

Comment:
I'm not sure the compiler can elide the second variant type check. Maybe `get_if`?

Reply:
Good point